lib/watcher-middleware.js

```diff
@@ -0,0 +1,56 @@
+'use strict';
+
+var url = require('url');
+var path = require('path');
+
+var setResponseHeaders = require('./utils/response-header');
+var errorHandler = require('./utils/error-handler');
+
+module.exports = function watcherMiddleware(watcher, options) {
+  options = options || {};
+
+  if (!options.hasOwnProperty('autoIndex')) {
+    // set autoIndex to be true if not provided
+    options.autoIndex = true;
+  }
+
+  return function(request, response, next) {
+    watcher.then(function(hash) {
+      var outputPath = path.normalize(hash.directory);
+
+      // set the x-broccoli header containing per request info used by the broccoli-middleware
+      var urlToBeServed = request.url;
+      var urlObj = url.parse(urlToBeServed);
+      var filename = path.join(outputPath, decodeURIComponent(urlObj.pathname));
+      var broccoliInfo = {
+        'url': urlToBeServed,
+        'outputPath': outputPath,
+        'filename': filename,
+      };
+      request.headers['x-broccoli'] = broccoliInfo;
+
+      var updatedFileName = setResponseHeaders(request, response, next, {
```

Comment:
1. I am not adding to set any extra headers. These headers were also set previously. I think the comment was more of an example to explain the workflow. The reason file specific headers are set in the middleware is because we want to provide these as the default implementation and let anyone override it downstream if needed rather than anyone who wants to insert itself before the asset serving middleware to almost copy over those header settings. 

2. I thought about splitting it but it wasn't very clean since it has dependencies on `stat` object and other things. I could do a small refactor such that it doesn't take `next`.